### PR TITLE
Add Tools > BaseFind + command line options

### PR DESCRIPTION
**Pull request by wargio**
# squash me

**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://cutter.re/docs/contributing/code/getting-started.html) to this repository
- [ ] I made sure to follow the project's [coding style](https://cutter.re/docs/contributing/code/development-guidelines.html)
- [ ] I've updated the [documentation](https://cutter.re/docs/user-docs.html) with the relevant information (if needed)

**Detailed description**

Adds Basefind to cutter (requires https://github.com/rizinorg/rizin/pull/3598).
Also adds various commandline features:

```
$ ./cutter -h
Usage: ./cutter [options] filename
A Qt and C++ GUI for rizin reverse engineering framework

Options:
  -h, --help                    Displays help on commandline options.
  --help-all                    Displays help including Qt specific options.
  -v, --version                 Displays version information.
  -A, --analysis <level>        Automatically open file and optionally start
                                analysis. Needs filename to be specified. May be
                                a value between 0 and 2: 0 = no analysis, 1 =
                                aaa, 2 = aaaa (experimental)
  -a, --arch <arch>             Sets a specific arch name
  -b, --bits <bits>             Sets a specific arch bits
  -c, --cpu <cpu>               Sets a specific cpu
  -o, --os <os>                 Sets a specific os
  -F, --format <name>           Force using a specific file format (bin plugin)
  -B, --base <base address>     Load binary at a specific base address
  -m, --map <map address>       Map the binary at a specific address
  -i <file>                     Run script file
  -p, --project <project file>  Load project file
  -w, --writemode               Open file in write mode
  -P, --phymode                 Disables virtual addressing
  --pythonhome <PYTHONHOME>     PYTHONHOME to use for embedded python
                                interpreter
  --no-output-redirect          Disable output redirection. Some of the output
                                in console widget will not be visible. Use this
                                option when debuging a crash or freeze and
                                output  redirection is causing some messages to
                                be lost.
  --no-plugins                  Do not load plugins
  --no-cutter-plugins           Do not load Cutter plugins
  --no-rizin-plugins            Do not load rizin plugins

Arguments:
  filename                      Filename to open.
```

image

image

image

### CODE REVIEW

File: src/core/Basefind.cpp [hunk 1]
```diff
@@ -0,0 +1,109 @@
+#include "Basefind.h"
```
Comment (XVilka) [inline src/core/Basefind.cpp:1]: Copyright?

File: src/core/Cutter.cpp [hunk 2]
```diff
@@ -4577,3 +4577,9 @@ void CutterCore::writeGraphvizGraphToFile(QString path, QString format, RzCoreGr
         }
     }
 }
+
+bool CutterCore::rebaseBin(RVA base_address)
+{
+    CORE_LOCK();
+    return rz_core_bin_rebase(core, base_address);
```
Comment (XVilka) [inline src/core/Cutter.cpp:4584]: Does it work now?

File: src/core/MainWindow.cpp [hunk 3]
```diff
@@ -1774,12 +1783,10 @@ void MainWindow::on_actionExport_as_code_triggered()
         return;
     }
 
-
-
     auto string = fromOwned(
-        dialog.selectedNameFilter() != instructionsInComments
-                ? rz_lang_byte_array(buffer.data(), size, typMap[dialog.selectedNameFilter()])
-                : rz_core_print_bytes_with_inst(rc, buffer.data(), 0, size));
+            dialog.selectedNameFilter() != instructionsInComments
```
Comment (XVilka) [inline src/core/MainWindow.cpp:1787]: Why this change?

File: src/tools/basefind/BaseFindDialog.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+#define BASEFIND_DIALOG_H
+
+#include <QDialog>
+#include <QListWidgetItem>
+#include <memory>
+
+#include <core/Cutter.h>
+
+namespace Ui {
+class BaseFindDialog;
```
Comment (XVilka) [inline src/tools/basefind/BaseFindDialog.h:11]: broken indentation?

**Review by XVilka — CHANGES_REQUESTED** [review]
The feature itself works, but when I choose "Set base address at ..." the functions window and disassembly didn't update, it seems. Even Dashboard didn't update:

Screenshot 2023-06-23 at 20 12 21

Opened `rizin-testbins/firmware/stm32f103-dapboot-v1.20-bluepill.bin`

File: src/core/MainWindow.cpp (hunk 3 above)
Comment (wargio) [reply src/core/MainWindow.cpp:1787]: that is clang format. let me revert it.

File: src/tools/basefind/BaseFindDialog.h (hunk 4 above)
Comment (wargio) [reply src/tools/basefind/BaseFindDialog.h:11]: i always run clang format before compiling, so it is "well formatted"

File: src/core/MainWindow.cpp (hunk 3 above)
Comment (wargio) [reply src/core/MainWindow.cpp:1787]: looks like correct for clang-format.

File: src/core/Cutter.cpp (hunk 2 above)
Comment (wargio) [reply src/core/Cutter.cpp:4584]: the other scenario is to open to a new window.

File: src/core/Basefind.cpp (hunk 1 above)
Comment (wargio) [reply src/core/Basefind.cpp:1]: the other files doesn't have it.

**Review by XVilka — APPROVED** [review]

**Comment by ITAYC0HEN** [comment]
nice! Update the docs where appropriate:
For example: https://github.com/rizinorg/cutter/blob/dev/docs/source/user-docs/command-line.rst

File: src/CutterApplication.cpp [hunk 5]
```diff
@@ -315,6 +385,22 @@ bool CutterApplication::parseCommandLineOptions()
             QObject::tr("level"));
     cmd_parser.addOption(analOption);
 
+    QCommandLineOption archOption({ "a", "arch" }, QObject::tr("Sets a specific arch name"),
```
Comment (ITAYC0HEN) [inline src/CutterApplication.cpp]:
```suggestion
    QCommandLineOption archOption({ "a", "arch" }, QObject::tr("Sets a specific architecture name"),
```

File: src/CutterApplication.cpp [hunk 6]
```diff
@@ -315,6 +385,22 @@ bool CutterApplication::parseCommandLineOptions()
...
+                                  QObject::tr("arch"));
+    cmd_parser.addOption(archOption);
+
+    QCommandLineOption bitsOption({ "b", "bits" }, QObject::tr("Sets a specific arch bits"),
```
Comment (ITAYC0HEN) [inline src/CutterApplication.cpp]:
```suggestion
    QCommandLineOption bitsOption({ "b", "bits" }, QObject::tr("Sets a specific architecture bits"),
```

File: src/CutterApplication.cpp [hunk 7]
```diff
@@ -315,6 +385,22 @@ bool CutterApplication::parseCommandLineOptions()
...
+                                  QObject::tr("bits"));
+    cmd_parser.addOption(bitsOption);
+
+    QCommandLineOption cpuOption({ "c", "cpu" }, QObject::tr("Sets a specific cpu"),
```
Comment (ITAYC0HEN) [inline src/CutterApplication.cpp]:
```suggestion
    QCommandLineOption cpuOption({ "c", "cpu" }, QObject::tr("Sets a specific CPU"),
```

File: src/CutterApplication.cpp [hunk 8]
```diff
@@ -315,6 +385,22 @@ bool CutterApplication::parseCommandLineOptions()
...
+                                 QObject::tr("cpu"));
+    cmd_parser.addOption(cpuOption);
+
+    QCommandLineOption osOption({ "o", "os" }, QObject::tr("Sets a specific os"),
```
Comment (ITAYC0HEN) [inline src/CutterApplication.cpp]:
```suggestion
    QCommandLineOption osOption({ "o", "os" }, QObject::tr("Sets a specific operating system"),
```

**Review by ITAYC0HEN — CHANGES_REQUESTED** [review]

File: src/tools/basefind/BaseFindDialog.ui [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+     <item row="1" column="0">
+      <widget class="QLabel" name="pointerSizeLabel">
+       <property name="sizePolicy">
+        <sizepolicy hsizetype="Preferred" vsizetype="Preferred">
+         <horstretch>0</horstretch>
+         <verstretch>0</verstretch>
+        </sizepolicy>
+       </property>
+       <property name="text">
+        <string>Pointer Size</string>
```
Comment (ITAYC0HEN) [inline src/tools/basefind/BaseFindDialog.ui]:
```suggestion
        <string>Pointer Size:</string>
```

File: src/tools/basefind/BaseFindDialog.ui [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+        <property name="text">
+         <string>64</string>
+        </property>
+       </item>
+      </widget>
+     </item>
+     <item row="2" column="0">
+      <widget class="QLabel" name="startAddressLabel">
+       <property name="text">
+        <string>Start Address</string>
```
Comment (ITAYC0HEN) [inline src/tools/basefind/BaseFindDialog.ui]:
```suggestion
        <string>Start Address:</string>
```

**Review by ITAYC0HEN — CHANGES_REQUESTED** [review]

**Comment by ITAYC0HEN** [comment]
Is "reopening" cutter a mandatory for the changes to take place? Is there currently no way to force the rebasing without re-opening?

**Comment by wargio** [comment]
> Is "reopening" cutter a mandatory for the changes to take place? Is there currently no way to force the rebasing without re-opening?

unfortunately for now yes.

**Comment by XVilka** [comment]
> The feature itself works, but when I choose "Set base address at ..." the functions window and disassembly didn't update, it seems. Even Dashboard didn't update:
> 
> Screenshot 2023-06-23 at 20 12 21
> Opened `rizin-testbins/firmware/stm32f103-dapboot-v1.20-bluepill.bin`

If I use "Reopen Cutter with with map address at 0x8000000" menu option (same file), it's even worse:

Screenshot 2023-06-25 at 10 52 40

Screenshot 2023-06-25 at 10 53 17

**Review by XVilka — APPROVED** [review]

**Pull request merged**